Detectors/EMCAL/workflow/src/PublisherSpec.cxx

```diff
@@ -0,0 +1,134 @@
+      constexpr auto persistency = o2::framework::Lifetime::Timeframe;
+      o2::header::DataHeader::SubSpecificationType subSpec = 0;
+      if (propagateMC) {
+        processAttributes->reader = std::make_shared<o2::framework::RootTreeReader>(treename.c_str(), // tree name
+                                                                                    filename.c_str(), // input file name
+                                                                                    nofEvents,        // number of entries to publish
+                                                                                    publishingMode,
+                                                                                    o2::framework::Output{ dto.origin, dto.description, subSpec, persistency },
+                                                                                    dtbrName.c_str(), // name of cluster branch
+                                                                                    o2::framework::Output{ mco.origin, mco.description, subSpec, persistency },
+                                                                                    mcbrName.c_str() // name of mc label branch
+        );
+      } else {
+        processAttributes->reader = std::make_shared<o2::framework::RootTreeReader>(treename.c_str(), // tree name
+                                                                                    filename.c_str(), // input file name
+                                                                                    nofEvents,        // number of entries to publish
+                                                                                    publishingMode,
+                                                                                    o2::framework::Output{ dto.origin, dto.description, subSpec, persistency },
+                                                                                    dtbrName.c_str() // name of cluster branch
+        );
+      }
+    }
+
+    auto processFunction = [processAttributes, propagateMC](o2::framework::ProcessingContext& pc) {
+      if (processAttributes->finished) {
+        return;
+      }
+
+      auto publish = [&processAttributes, &pc, propagateMC]() {
+        o2::emcal::EMCALBlockHeader emcheader(true);
+        if (processAttributes->reader->next()) {
+          (*processAttributes->reader)(pc, emcheader);
+        } else {
+          processAttributes->reader.reset();
+          return false;
+        }
+        return true;
+      };
+
+      bool active(true);
+      if (!publish()) {
+        active = false;
+        // Send dummy header with no payload option
+        o2::emcal::EMCALBlockHeader dummyheader(false);
+        pc.outputs().snapshot(o2::framework::OutputRef{ "output", 0, { dummyheader } }, 0);
+        if (propagateMC) {
+          pc.outputs().snapshot(o2::framework::OutputRef{ "outputMC", 0, { dummyheader } }, 0);
+        }
+      }
```

Comment:
If you drop the handling of the no-more-data case, you avoid the exception in the receiving process.